compiler/z/codegen/OMRRegisterDependency.cpp

```diff
@@ -680,167 +680,269 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
-   TR::CodeGenerator * cg)
+TR_S390RegisterDependencyGroup::genBitVectOfAssignableGPRs(TR::CodeGenerator * cg,
+                                                           uint32_t numberOfRegisters)
    {
-   int32_t i;
    uint32_t availRegMap = cg->machine()->genBitVectOfAssignableGPRs();
-   TR::Register * virtReg;
-   TR::RealRegister::RegNum realReg;
 
-   for (i = 0; i < numberOfRegisters; i++)
+   for (uint32_t i = 0; i < numberOfRegisters; i++)
       {
-      realReg = _dependencies[i].getRealRegister();
+      TR::RealRegister::RegNum realReg = _dependencies[i].getRealRegister();
       if (TR::RealRegister::isGPR(realReg))
-         {
          availRegMap &= (~TR::RealRegister::getBitMask(realReg));
-         }
       }
 
    return availRegMap;
    }
 
-uint32_t
-TR_S390RegisterDependencyGroup::checkDependencyGroup(TR::Instruction * currentInstruction, uint32_t          numberOfRegisters,
-   TR::CodeGenerator * cg)
-   {
-   TR::Compilation *comp = cg->comp();
-   int32_t i, j;
-   int32_t numReqPairs = 0;
-   int32_t numAvailPairs = 0;
-   TR::Register * virtReg;
-   TR::RealRegister::RegNum realReg;
 
-   uint32_t availRegMap = genBitVectOfAssignableGPRs(currentInstruction, numberOfRegisters, cg);
-   cg->setAvailableHPRSpillMask(0xFFFF0000);
+/** \brief
+ *     Checks to ensure that we have enough register pairs available to satisfy all register
+ *     dependencies in the current denepency group and aborts the
+ *     compilation if not enough register pairs are available.
+ *
+ *     Also, if highword facility is supported, this function checks the type of each register
+ *     dependency condition and sets HPR related flags for future register assignment steps.
+ *
+ *  \param cg
+ *     The code generator used to query various machine state.
+ *
+ *  \param currentInstruction
+ *     The instruction on which the register dependency group is attached.
+ *
+ *  \param availableGPRMap
+ *     The bitmask of avilable GRP registers.
+ *
+ *  \param numOfDependencies
+ *     The number of pre or post dependencies in the register dependency group.
+ **/
+void
+TR_S390RegisterDependencyGroup::checkRegisterPairSufficiencyAndHPRAssignment(TR::CodeGenerator *cg,
+                                                             TR::Instruction* currentInstruction,
+                                                             const uint32_t availableGPRMap,
+                                                             uint32_t numOfDependencies)
+   {
+   bool isHighWordSupported = cg->supportsHighWordFacility() && !(cg->comp()->getOption(TR_DisableHighWordRA));
 
-   // Count the number of pairs remaining after deps are filled
-   for (i = 0; i < NUM_S390_GPR; i += 2)
-      {
-      if ((availRegMap >> i) & 0x3 == 0x3)
-         {
-         numAvailPairs++;
-         }
-      }
+   uint32_t numReqPairs = 0, numAvailPairs = 0;
+   int32_t i;
```

Comment:
This declaration can be pushed into the for loop (thus reducing the scope of the variable and making a more idiomatic looking loop). Similarly elsewhere in the commit.